### PR TITLE
Fix bug #13618: too restrictive recognition of homogeneous relations

### LINKED ISSUE #13618: Setoid rewriting on Type
#### Description of the problem

I ran into an issue with setoid rewriting for equivalences on Type, see below for a minimal example. There is a related [Zulip topic](https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/Issue.20with.20setoid.20rewriting.20on.20Type).

If I just assume an equivalence respecting sum types, rewriting below a sum fails although it can be easily be done by hand. It seems this can be fixed by explicitly aliasing `sum` with full type annotation and annotating occurring data types as well.

```coq
Require Import Morphisms Setoid.

Section Equiv.

  Variable equiv : Type -> Type -> Prop.
  Hypothesis equiv_equiv : Equivalence equiv.
  Hypothesis equiv_sum : Proper (equiv ==> equiv ==> equiv) sum.

  Goal forall X, equiv X bool -> equiv (X + X) X.
  Proof.
    intros X H. Fail rewrite H.
    (* The command has indeed failed with message: *)
    (* Cannot find an homogeneous relation to rewrite. *)
    eapply Equivalence_Transitive; try apply equiv_sum; try apply H; try apply Equivalence_Reflexive.
  Abort.

  Definition sum' : Type -> Type -> Type := sum.

  Hypothesis equiv_sum' : Proper (equiv ==> equiv ==> equiv) sum'.

  Goal forall X, equiv X (bool : Type) -> equiv (sum' X X) X.
  Proof.
    intros X H. rewrite H.
  Abort.

End Equiv.
```

Using `CMorphism` produces the same error and @mattam82 posted another not fully satisfactory workaround on Zulip.

#### Coq Version

Coq 8.12.0

### PR DESCRIPTION
**Pull request by mattam82**
**Kind:** bug fix

We were looking at the wrong types to determine if we were passed an homogeneous relation to rewrite with.
This should allow to easily rewrite with *homogeneous* type equivalence relations (comparing two types in the same universe). As Coq has cumulativity, heterogeneous type equivalences can always be embedded into homogeneous ones, by lifting the equivalence in the highest of the two universes.

Fixes / closes #13618

- [x] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: plugins/ltac/rewrite.ml [hunk 1]
```diff
@@ -494,7 +494,15 @@ let decompose_app_rel env evd t =
...
   let ty = try Retyping.get_type_of ~lax:true env evd rel with Retyping.RetypeError _ -> error_no_relation () in
   let () = if not (Reductionops.is_arity env evd ty) then error_no_relation () in
-  (rel, t1, t2)
+  let ty1, ty2, concl =
+    match Reductionops.splay_arity env evd ty with
+    | [_, ty1; _, ty2], concl ->
+      if noccurn evd 0 ty2 then
+        ty1, subst1 mkProp ty2, concl
+      else error_no_relation ()
+    | _ -> assert false
```
Comment (SkySkimmer) [inline plugins/ltac/rewrite.ml:503]: Why is this branch impossible?

Comment (mattam82) [reply plugins/ltac/rewrite.ml:503]: Because of the two tests above, I think

**Comment by mattam82** [comment]
@SkySkimmer ?

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**